src/core/TESTING.md

```diff
@@ -1087,3 +1095,275 @@ describe('Plugin', () => {
+the interval of 100ms was respected. Even using a subscription based test to try to do so would result
+in potential flakiness, as the subscription execution could trigger on the `101ms` time frame for example.
+
+It also may be important to note:
+- as we need to convert the observable to a promise and wait for the result, the test is `async`
+- we need to perform observable transformation (`take` + `toArray`) in the test to have an usable value to assert against.
+
+Marble testing would allow to get rid of these limitations. An equivalent and improved marble test could be:
+
+```typescript
+  describe('getUpdate$', () => {
+    it('emits updates every 100ms', () => {
+      getTestScheduler().run(({ expectObservable }) => {
+        const { getUpdate$ } = service.setup();
+        expectObservable(getUpdate$(), '301ms !').toBe('100ms a 99ms b 99ms c', {
+          a: 'update-1',
+          b: 'update-2',
+          c: 'update-3',
+        });
+      });
+    });
+  });
+```
+
+Notes:
+- the test is now synchronous
+- the second parameter of `expectObservable` (`'301ms !'`) is used to perform manual unsubscription to the observable, as 
+  `interval` never ends.
+- an emission is considered a time frame, meaning that after the initial `a` emission, we are at the frame `101`, not `100`
+  which is why we are then only using a `99ms` gap between a->b and b->c.
+
+#### Testing observable completion
+
+Let's 'improve' our `getUpdate$` API by allowing the consumer to manually terminate the observable chain using
+a new `abort$` option:
+
+```typescript
+class FooService {
+  setup() {
+    return {
+      getUpdate$: ({ abort$ = EMPTY }: { abort$?: Observable<undefined> } = {}) => {
+        return interval(100).pipe(
+            takeUntil(abort$), 
+            map((count) => `update-${count + 1}`)
+        );
+      },
+    };
+  }
+}
+```
+
+We would then add a test to assert than this new option usage is respected:
+
+```typescript
+it('getUpdate$ completes when `abort$` emits', () => {
+  const service = new FooService();
+  getTestScheduler().run(({ expectObservable, hot }) => {
+    const { getUpdate$ } = service.setup();
+    const abort$ = hot('149ms a', { a: undefined });
+    expectObservable(getUpdate$({ abort$ })).toBe('100ms a 48ms |', {
+      a: 'update-1',
+    });
+  });
+});
+```
+
+Notes:
+  - the `|` symbol represents the completion of the observable.
+  - we are here using the `hot` testing utility to create the `abort$` observable to ensure correct emission timing.
+ 
+#### Testing observable errors 
+
+Testing errors thrown by the observable is very close to the previous examples and is done using
+the third parameter of `expectObservable`.
+
+Say we have a service in charge of processing data from an observable and returning the results in a new observable:
+
+```typescript
+interface SomeDataType {
+  id: string;
+}
+
+class ProcessError extends Error {
```

Comment:
I like we promote domain-specific error usage. OTOH, it rather distracts from the code sample.

Reply:
I mostly did it because the error comparison is dumb when using default errors.

If using `throw new Error(`invalid data: '${data.id}'`);`

```ts
expectObservable(processDataStream(data)).toBe(
      '--a--#',
      {
        a: { id: 'a', processed: 'additional-data' },
      },
      new ProcessError(`invalid data: 'invalid'`)
    );
```

would become

```ts
expectObservable(processDataStream(data)).toBe(
      '--a--#',
      {
        a: { id: 'a', processed: 'additional-data' },
      },
      `'[Error: invalid data: 'invalid']'`
    );
```

But I agree that it does increase and complexify the code sample. I'm fine either way

Reply:
I'd rather keep it simple, but it's up to you